internal/moduletest/hcl/variable_cache.go

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package hcl
+
+import (
+	"github.com/hashicorp/hcl/v2"
+	"github.com/zclconf/go-cty/cty"
+
+	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/backend/backendrun"
+	"github.com/hashicorp/terraform/internal/configs"
+	"github.com/hashicorp/terraform/internal/lang/langrefs"
+	"github.com/hashicorp/terraform/internal/terraform"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+)
+
+// VariableCaches contains a mapping between test run blocks and evaluated
+// variables. This is used to cache the results of evaluating variables so that
+// they are only evaluated once per run.
+//
+// Each run block has its own configuration and therefore its own set of
+// evaluated variables.
+type VariableCaches struct {
+	GlobalVariables map[string]backendrun.UnparsedVariableValue
+	FileVariables   map[string]hcl.Expression
+
+	caches map[string]*VariableCache
+}
+
+// VariableCache contains the cache for a single run block. This cache contains
+// the evaluated values for global and file-level variables.
+type VariableCache struct {
+	config *configs.Config
+
+	globals terraform.InputValues
+	files   terraform.InputValues
+
+	values *VariableCaches // back reference so we can access the stored values
+}
+
+// GetCache returns the cache for the named run. If the cache does not exist, it
+// is created and returned.
+func (caches *VariableCaches) GetCache(name string, config *configs.Config) *VariableCache {
+	if caches.caches == nil {
+		caches.caches = make(map[string]*VariableCache)
+	}
+
+	cache, exists := caches.caches[name]
+	if !exists {
+		cache = &VariableCache{
+			config:  config,
+			globals: make(terraform.InputValues),
+			files:   make(terraform.InputValues),
+			values:  caches,
+		}
+		caches.caches[name] = cache
+	}
+	return cache
+}
+
+// GetGlobalVariable returns a value for the named global variable evaluated
+// against the named run.
+//
+// This function caches the result of evaluating the variable so that it is
+// only evaluated once per run.
+func (cache *VariableCache) GetGlobalVariable(name string) (*terraform.InputValue, tfdiags.Diagnostics) {
+	val, exists := cache.globals[name]
+	if exists {
+		return val, nil
+	}
+
+	variable, exists := cache.values.GlobalVariables[name]
+	if !exists {
+		return nil, nil
+	}
+
+	// TODO: We should also introduce a way to specify the mode in the test
+	//   file itself. Suggestion, optional variable blocks.
+	parsingMode := configs.VariableParseHCL
+
+	if cfg, exists := cache.config.Module.Variables[name]; exists {
+		parsingMode = cfg.ParsingMode
+	}
+
+	value, diags := variable.ParseVariableValue(parsingMode)
+	if diags.HasErrors() {
+		// In this case, the variable exists but we couldn't parse it. We'll
```

Comment:
Recommend adding to the function comments for `GetGlobalVariable` and `GetFileVariable` to say what gets returned if the variable can't be parsed or evaluated and remind callers to check diags before trying to use the value.

Reply:
Done!